src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -308,8 +308,17 @@ private DecisionTree AddByValue(DecisionTree.ByType byType, BoundConstantPattern
             if (forType == null)
             {
                 var type = value.Value.Type;
-                var narrowedExpression = GetBoundPatternMatchingLocal(type);
-                forType = new DecisionTree.ByValue(narrowedExpression, type.TupleUnderlyingTypeOrSelf(), narrowedExpression.LocalSymbol);
+                if (byType.Type.Equals(type, TypeCompareKind.AllIgnoreOptions))
+                {
+                    // reuse the input expression when we have an equivalent type to reduce the number of generated temps
+                    forType = new DecisionTree.ByValue(byType.Expression, type.TupleUnderlyingTypeOrSelf(), null);
```

Comment:
Yes, but that parameter is non-null only when it represents a (possibly) newly allocated temp (so we can keep track of which temps exist in the final decision tree). We are reusing the expression, so we are not allocating a new temp at this node.